presto-main/src/main/java/com/facebook/presto/metadata/Split.java

```diff
@@ -77,6 +83,17 @@ public Lifespan getLifespan()
         return lifespan;
     }
 
+    @JsonProperty
+    public SplitContext getSplitContext()
+    {
+        return splitContext;
+    }
+
+    public void setSplitContext(SplitContext splitContext)
```

Comment:
You still need `equals` and `hashCode` function, so that `splitContext` can be immutable.

Reply:
So I have these two version of code both ready
1.Provide SplitContext as a mutable field, as the code above
2.Use hashCode and equals to indicate if they're the same Split (not include SplitContext) 
https://github.com/kewang1024/presto/commit/376fb8a229021a6a70213026b5bd38169339bc97
(https://github.com/kewang1024/presto/commits/test_pass_cache)

When ``SourcePartitionedScheduler`` use ``SimpleNodeSelector`` to do scheduling for pending task, we may change some split's SplitContext to cacheable and return those split back to ``SourcePartitionedScheduler`` 
``SourcePartitionedScheduler`` will evict those split from pendingTask (a split Set)
for 
1. Pro: more straightforward  since we are removing the same split from split set.
  Con: Split not immutable any more
2.Pro: Split immutable
  Con: seems a little hacky to evict from a set using a different Split (but they're same because we don't include SplitContext in ``equals`` and ``hash`` method), and possibly cause confusion if we lose context at some point

Also want to know your opinion on it @shixuan-fan @wenleix

Reply:
We don't use mutable data structure in our codebase. Otherwise, we can avoid creating `SplitContext` by directly change the context for `HiveSplit`

Reply:
I agree that an immutable class is easier to reason and understand.